src/raftstore/store/fsm/apply.rs

```diff
@@ -3917,4 +3919,126 @@ mod tests {
+
+        router.schedule_task(1, Msg::Registration(reg));
+
+        let mut reg = Registration::default();
+        reg.id = 2;
+        reg.term = 2;
+        reg.region.set_id(2);
+        reg.region.set_start_key(b"l5".to_vec());
+        reg.region.set_end_key(b"z5".to_vec());
+        reg.apply_state.set_applied_index(3);
+        reg.region.mut_region_epoch().set_version(3);
+        reg.region.mut_peers().push(new_peer(1, 2));
+        reg.region.mut_region_epoch().set_conf_ver(1);
+
+        router.schedule_task(2, Msg::Registration(reg));
+
+        let (send, rec) = channel();
+
+        let mut apply0_handle = 0;
+        let mut apply1_handle = 0;
+
+        let mut worker: Vec<JoinHandle<()>> = Vec::new();
+        for _i in 0..1 {
+            let tx1 = send.clone();
+            let router = router.clone();
+            let t = thread::spawn(move || {
+                for _i in 0..1000000 {
+                    let sender = tx1.clone();
+                    router.schedule_task(
+                        1,
+                        Msg::Validate(
+                            1,
+                            Box::new(move |_| match thread::current().name() {
+                                Some("apply-0::test_apply_tilt") => sender.send(0).unwrap(),
+                                _ => sender.send(1).unwrap(),
+                            }),
+                        ),
+                    );
+                }
+            });
+            worker.push(t);
+        }
+
+        for _i in 0..1 {
+            let tx1 = send.clone();
+            let router = router.clone();
+            let t = thread::spawn(move || {
+                for _i in 0..1000000 {
+                    let sender = tx1.clone();
+                    router.schedule_task(
+                        2,
+                        Msg::Validate(
+                            2,
+                            Box::new(move |_| match thread::current().name() {
+                                Some("apply-0::test_apply_tilt") => sender.send(0).unwrap(),
+                                _ => sender.send(1).unwrap(),
+                            }),
+                        ),
+                    );
+                }
+            });
+            worker.push(t);
+        }
+
+        for w in worker {
+            w.join().unwrap();
+        }
+
+        for _i in 0..2000000 {
+            match rec.recv().unwrap() {
+                0 => apply0_handle += 1,
+                1 => apply1_handle += 1,
+                _ => {}
+            };
+        }
+        let sub = if apply0_handle > apply1_handle {
+            apply0_handle - apply1_handle
+        } else {
+            apply1_handle - apply0_handle
+        };
+
+        println!(
+            "apply0 handle {}  apply1 handle {}",
+            apply0_handle, apply1_handle
+        );
+        let sum = apply0_handle + apply1_handle;
+        println!("sub is {} sum is {}", sub, sum);
```

Comment:
Is there any expectation of the `sub`? Is there a threshold of the `sub` we should trigger the test failure?

Reply:
When recurring, sub should be 2000000 to indicate that the problem exists, but we have fixed the problem, then sub should not equal 2000000. We hope that the sub is as small as possible.

Reply:
But in extreme cases, sub may be 2000000, and I can't guarantee that this is not the case, although the probability is very small, but it will happen in theory.

Reply:
I think the `println!` is useless because the CI has no idea to determine is it expected base on the output, maybe we should add a comment (e.g: `FIXME:  should define the `sub` threshold of CI failure`).

Reply:
OK